ludwig/trainers/trainer.py

```diff
@@ -1189,8 +1183,26 @@ def _train_loop(
                     if self.is_coordinator():
                         progress_tracker.save(os.path.join(save_path, TRAINING_PROGRESS_TRACKER_FILE_NAME))
 
+                # If this was the last batch, then increment the epoch counter and invoke the `on_epoch_end` callback.
+                if not has_called_on_epoch_end and batcher.last_batch():
+                    progress_tracker.epoch += 1
+                    self.callback(lambda c: c.on_epoch_end(self, progress_tracker, save_path))
+
+                    # Ensures that we only call `on_epoch_end` once, for example, if evaluation is configured to run
+                    # several times per epoch.
+                    has_called_on_epoch_end = True
+
                 if should_break:
-                    return should_break
+                    return True
```

Comment:
The early return here seems odd given that the only following piece of code is an epoch increment and callback that won't happen before the end of the epoch. What if we set `should_break = False` at the top of this method, then returned `should_break` at the end? That would let us remove the duplicate block above, remove this block, and make it so we don't need the `has_called_on_epoch_end` flag.

Reply:
Ah, that's a brain fart on my end -- thanks for the suggestion. Returning `should_break` is definitely simpler!